### PR TITLE
ledger rewrite

### PR DESCRIPTION
**Pull request by xiegeo**
## Description
#6 Optimize Ledger Performance

## Motivation and Context
Ledger was slow and does not scale with lots of data.

Currently, this pr should be more feature complete than then before the rewrite.

## How to test this PR?
TODO

TODOs
- [x] Rewrite object upload.
- [x] Rewrite multipart upload.
- [ ] Test coverage

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
Support for some APIs was improved.
- [x] Breaking change (fix or feature that would cause existing functionality to change)
Breaks database compatibility.

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [x] Documentation needed
- [x] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

File: cmd/gateway/s3x/s3.proto [hunk 1]
```diff
@@ -28,29 +28,20 @@ message InfoResponse {
 // for keeping track of buckets, objects, and their corresponding IPFS hashes
 message Ledger {
     // key = bucket name
-    map<string, LedgerBucketEntry> buckets = 1 [(gogoproto.nullable) = false];
+    map<string, LedgerBucketEntry> buckets = 1;
     // key = partID
-    map<string, MultipartUpload> multipartUploads = 2 [(gogoproto.nullable) = false];
```
Comment (bonedaddy) [inline cmd/gateway/s3x/s3.proto:33]: We should keep the nullable option since it removes the pointer and will allow us to unmarshal/marshal the data faster

File: cmd/gateway/s3x/ipfs.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+func ipfsBucket(ctx context.Context, dag pb.NodeAPIClient, h string) (*Bucket, error) {
+	b := &Bucket{}
+	if err := ipfsUnmarshal(ctx, dag, h, b); err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+// bucketToIPFS takes a bucket and stores it on IPFS
+func ipfsSave(ctx context.Context, dag pb.NodeAPIClient, m marshaller) (string, error) {
```
Comment (bonedaddy) [inline cmd/gateway/s3x/ipfs.go:54]: Update comment

File: cmd/gateway/s3x/ipfs.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	data, err := m.Marshal()
+	if err != nil {
+		return "", err
+	}
+	return ipfsSaveBytes(ctx, dag, data)
+}
+
+// ipfsBytes returns data from IPFS using its hash
+func ipfsSaveBytes(ctx context.Context, dag pb.NodeAPIClient, data []byte) (string, error) {
```
Comment (bonedaddy) [inline cmd/gateway/s3x/ipfs.go:63]: Update comment here as well

File: cmd/gateway/s3x/s3.proto [hunk 4]
```diff
@@ -28,29 +28,20 @@ message InfoResponse {
...
-// LedgerObjectEntry is an individual entry within an bucket, containing information about an object
-message LedgerObjectEntry {
-    string ipfsHash = 1;
-    string name = 2;
-}
-
 // BucketInfo is used to store s3 bucket metadata
 message BucketInfo {
-    // name is the name of the bucket
-    string name = 1;
```
Comment (bonedaddy) [inline cmd/gateway/s3x/s3.proto]: Why is the name being removed? Without the name here, and only having the ipfs hash of the bucket, you can't retrieve the name of the bucket.

File: cmd/gateway/s3x/ipfs.go (hunk 2 above)
Comment (xiegeo) [reply cmd/gateway/s3x/ipfs.go:54]: good catch

File: cmd/gateway/s3x/ipfs.go (hunk 3 above)
Comment (xiegeo) [reply cmd/gateway/s3x/ipfs.go:63]: nice catch

File: cmd/gateway/s3x/s3.proto (hunk 4 above)
Comment (xiegeo) [reply cmd/gateway/s3x/s3.proto]: ok, I'll add this back.

**Comment by xiegeo** [comment]
currently failing because of spellcheck localization

``` bash
xiege@LAPTOP-NPU545OC MINGW64 ~/go/src/github.com/RTradeLtd/s3x/cmd/gateway/s3x (ledger-rewrite)
$ misspell .

xiege@LAPTOP-NPU545OC MINGW64 ~/go/src/github.com/RTradeLtd/s3x/cmd/gateway/s3x (ledger-rewrite)
$ misspell -locale US .
gateway-s3x-ledger-store.go:34:72: "optimise" is a misspelling of "optimize"
```

Going to apply the spell fix to make it upstreamable.

File: cmd/gateway/s3x/s3.proto (hunk 1 above)
Comment (xiegeo) [reply cmd/gateway/s3x/s3.proto:33]: I thought more about this, I'm converting this away from saving as a single object for efficiency, and because there this no way to get a safe pointer from a map value, using a pointer is more convenient to code.

Comment (bonedaddy) [reply cmd/gateway/s3x/s3.proto:33]: Sounds good

File: cmd/gateway/s3x/s3.proto [hunk 5]
```diff
@@ -114,8 +107,8 @@ message ObjectPartInfo {
 }
 
 message MultipartUpload {
-    string bucket = 1;
-    string object = 2;
+    ObjectInfo objectInfo = 1;
     string id = 3;
```
Comment (bonedaddy) [inline cmd/gateway/s3x/s3.proto]: this should be `string id = 2` now that the `object` field was removed

File: cmd/gateway/s3x/s3.proto [hunk 6]
```diff
@@ -114,8 +107,8 @@ message ObjectPartInfo {
...
-    repeated ObjectPartInfo objectParts = 4 [(gogoproto.nullable) = false];
+    //map of index to parts
+    map<int64, ObjectPartInfo>  objectParts = 4 [(gogoproto.nullable) = false];
```
Comment (bonedaddy) [inline cmd/gateway/s3x/s3.proto]: this should be `map<int64, ObjectPartInfo>  objectParts = 3 [(gogoproto.nullable) = false];` now that the `object` field was removed

File: cmd/gateway/s3x/s3.proto [hunk 7]
```diff
@@ -101,10 +94,10 @@ message ObjectInfo {
 
 // ObjectPartInfo contains information an individual object client
 message ObjectPartInfo {
-    // this should be an "int" type in Golang
+    // convertable to "int" type in minio.PartInfo
     int64 number = 1;
     string name = 2;
-    string etag = 3;
+    //string etag = 3; also use dataHash
```
Comment (bonedaddy) [inline cmd/gateway/s3x/s3.proto]: If this field is no longer going to be used, we should decrease the field number. With protocol buffers you always want to make sure that you increment the field numbers `1, 2, 3, 4, 5, etc...` as it requires the fewest bytes to encode the information for the first `1->15` fields.

File: cmd/gateway/s3x/gateway-s3x-ledger-store.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+	}
+	_, err = ls.saveBucket(ctx, bucket, b.Bucket)
+	return missing, err
+	//todo: gc on ipfs
+}
+
+//PutObject saves an object by hash into the given bucket
+func (ls *ledgerStore) PutObject(ctx context.Context, bucket, object string, obj *Object) error {
+	defer ls.locker.write(bucket)()
+	return ls.putObject(ctx, bucket, object, obj)
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger-store.go:154]: We might want to do:
```
err := ls.putObject(ctx, bucket, object, obj)
ls.locker.wirte(bucket)()
return err
```

Would this `defer` technically run before `ls.putObject` is finished?

Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger-store.go:154]: Your change is a slight optimization, but otherwise the code is the same, unless putObject panics.

Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-ledger-store.go:154]: good point

**Comment by bonedaddy** [comment]
I think this test failure is something wrong with badger. Occasionally badger fucks up like this.

**Comment by xiegeo** [comment]
> 
> 
> I think this test failure is something wrong with badger. Occasionally badger fucks up like this.

<details> <summary>gateway-s3x-bucket_test.go:42: DB.Close: remove /home/travis/gopath/src/github.com/RTradeLtd/s3x/cmd/gateway/s3x/tmp-bucket-test/LOCK: no such file or directory</summary>
=== RUN   TestGateway_Bucket/DeleteBucket/Bucket2-NotFound

badger 2020/01/27 13:35:29 DEBUG: Storing value log head: {Fid:0 Len:29 Offset:184}

badger 2020/01/27 13:35:29 INFO: Got compaction priority: {level:0 score:1.73 dropPrefix:[]}

badger 2020/01/27 13:35:29 INFO: Running for level: 0

badger 2020/01/27 13:35:29 DEBUG: LOG Compact. Added 3 keys. Skipped 0 keys. Iteration took: 538.045µs

badger 2020/01/27 13:35:29 WARNING: LOG Compact FAILED with error: open tmp-bucket-test/000002.sst: no such file or directory

While opening new table: 2

github.com/dgraph-io/badger/v2.(*levelsController).compactBuildTables.func2

	/home/travis/gopath/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.1-0.20191128171928-407e5bde06c4/levels.go:596

github.com/dgraph-io/badger/v2.(*levelsController).compactBuildTables.func3

	/home/travis/gopath/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.1-0.20191128171928-407e5bde06c4/levels.go:620

runtime.goexit

	/home/travis/.gimme/versions/go1.13.6.linux.amd64/src/runtime/asm_amd64.s:1357

While running compaction for: {elog:0xc0002be000 thisLevel:0xc0000b2900 nextLevel:0xc0000b2960 top:[0xc00037e000] bot:[] thisRange:{left:[] right:[] inf:true} nextRange:{left:[33 98 97 100 103 101 114 33 104 101 97 100 0 0 0 0 0 0 0 0] right:[47 108 101 100 103 101 114 82 111 111 116 47 98 47 98 117 99 107 101 116 49 255 255 255 255 255 255 255 255] inf:false} thisSize:0 dropPrefix:[]}

github.com/dgraph-io/badger/v2.(*levelsController).compactBuildTables

	/home/travis/gopath/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.1-0.20191128171928-407e5bde06c4/levels.go:652

github.com/dgraph-io/badger/v2.(*levelsController).runCompactDef

	/home/travis/gopath/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.1-0.20191128171928-407e5bde06c4/levels.go:815

github.com/dgraph-io/badger/v2.(*levelsController).doCompact

	/home/travis/gopath/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.1-0.20191128171928-407e5bde06c4/levels.go:884

github.com/dgraph-io/badger/v2.(*DB).close

	/home/travis/gopath/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.1-0.20191128171928-407e5bde06c4/db.go:462

github.com/dgraph-io/badger/v2.(*DB).Close.func1

	/home/travis/gopath/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.1-0.20191128171928-407e5bde06c4/db.go:397

sync.(*Once).doSlow

	/home/travis/.gimme/versions/go1.13.6.linux.amd64/src/sync/once.go:66

sync.(*Once).Do

	/home/travis/.gimme/versions/go1.13.6.linux.amd64/src/sync/once.go:57

github.com/dgraph-io/badger/v2.(*DB).Close

	/home/travis/gopath/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.1-0.20191128171928-407e5bde06c4/db.go:396

github.com/RTradeLtd/go-ds-badger/v2.(*Datastore).Close

	/home/travis/gopath/pkg/mod/github.com/!r!trade!ltd/go-ds-badger/v2@v2.1.0/datastore.go:330

github.com/ipfs/go-datastore/keytransform.(*Datastore).Close

	/home/travis/gopath/pkg/mod/github.com/ipfs/go-datastore@v0.3.1/keytransform/keytransform.go:193

github.com/RTradeLtd/s3x/cmd/gateway/s3x.(*ledgerStore).Close

	/home/travis/gopath/src/github.com/RTradeLtd/s3x/cmd/gateway/s3x/gateway-s3x-ledger.go:23

github.com/RTradeLtd/s3x/cmd/gateway/s3x.(*xObjects).Shutdown

	/home/travis/gopath/src/github.com/RTradeLtd/s3x/cmd/gateway/s3x/gateway-s3x.go:193

github.com/RTradeLtd/s3x/cmd/gateway/s3x.TestGateway_Bucket.func1

	/home/travis/gopath/src/github.com/RTradeLtd/s3x/cmd/gateway/s3x/gateway-s3x-bucket_test.go:41

github.com/RTradeLtd/s3x/cmd/gateway/s3x.TestGateway_Bucket

	/home/travis/gopath/src/github.com/RTradeLtd/s3x/cmd/gateway/s3x/gateway-s3x-bucket_test.go:148

testing.tRunner

	/home/travis/.gimme/versions/go1.13.6.linux.amd64/src/testing/testing.go:909

runtime.goexit

	/home/travis/.gimme/versions/go1.13.6.linux.amd64/src/runtime/asm_amd64.s:1357: {elog:0xc0002be000 thisLevel:0xc0000b2900 nextLevel:0xc0000b2960 top:[0xc00037e000] bot:[] thisRange:{left:[] right:[] inf:true} nextRange:{left:[33 98 97 100 103 101 114 33 104 101 97 100 0 0 0 0 0 0 0 0] right:[47 108 101 100 103 101 114 82 111 111 116 47 98 47 98 117 99 107 101 116 49 255 255 255 255 255 255 255 255] inf:false} thisSize:0 dropPrefix:[]}

badger 2020/01/27 13:35:29 WARNING: While forcing compaction on level 0: While running compaction for: {elog:0xc0002be000 thisLevel:0xc0000b2900 nextLevel:0xc0000b2960 top:[0xc00037e000] bot:[] thisRange:{left:[] right:[] inf:true} nextRange:{left:[33 98 97 100 103 101 114 33 104 101 97 100 0 0 0 0 0 0 0 0] right:[47 108 101 100 103 101 114 82 111 111 116 47 98 47 98 117 99 107 101 116 49 255 255 255 255 255 255 255 255] inf:false} thisSize:0 dropPrefix:[]}: While opening new table: 2: open tmp-bucket-test/000002.sst: no such file or directory

--- FAIL: TestGateway_Bucket (0.42s)

    --- PASS: TestGateway_Bucket/MakeBucketWithLocation (0.38s)

        --- PASS: TestGateway_Bucket/MakeBucketWithLocation/Bucket1-Success (0.38s)

        --- PASS: TestGateway_Bucket/MakeBucketWithLocation/Bucket1-AlreadyExists (0.00s)

    --- PASS: TestGateway_Bucket/Bucket_Created_Time_Test (0.00s)

    --- PASS: TestGateway_Bucket/GetBucketInfo (0.00s)

        --- PASS: TestGateway_Bucket/GetBucketInfo/Bucket1-Found (0.00s)

        --- PASS: TestGateway_Bucket/GetBucketInfo/Bucket2-NotFound (0.00s)

    --- PASS: TestGateway_Bucket/ListBuckets (0.00s)

    --- PASS: TestGateway_Bucket/DeleteBucket (0.00s)

        --- PASS: TestGateway_Bucket/DeleteBucket/Bucket1-Found (0.00s)

        --- PASS: TestGateway_Bucket/DeleteBucket/Bucket2-NotFound (0.00s)

    gateway-s3x-bucket_test.go:42: DB.Close: remove /home/travis/gopath/src/github.com/RTradeLtd/s3x/cmd/gateway/s3x/tmp-bucket-test/LOCK: no such file or directory
</details>

File: cmd/gateway/s3x/s3.proto (hunk 7 above)
Comment (xiegeo) [reply cmd/gateway/s3x/s3.proto]: ok

File: cmd/gateway/s3x/s3.proto (hunk 6 above)
Comment (xiegeo) [reply cmd/gateway/s3x/s3.proto]: got it

File: cmd/gateway/s3x/s3.proto (hunk 5 above)
Comment (xiegeo) [reply cmd/gateway/s3x/s3.proto]: ok

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+func (ls *ledgerStore) AbortMultipartUpload(bucket, multipartID string) error {
+	err := ls.AssertBucketExits(bucket)
+	if err != nil {
+		return err
+	}
+	return ls.DeleteMultipartID(multipartID)
+}
+
+// NewMultipartUpload is used to store the initial start of a multipart upload request
+func (ls *ledgerStore) NewMultipartUpload(multipartID string, info *ObjectInfo) error {
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger-multipart.go:27]: Where's the complete multipart upload call?

File: cmd/gateway/s3x/gateway-s3x-object.go [hunk 10]
```diff
@@ -176,42 +144,41 @@ func (x *xObjects) PutObject(
...
-	}
-	// update internal ledger state with the new object
-	if err := x.ledgerStore.AddObjectToBucket(bucket, object, objectHash); err != nil {
-		return objInfo, x.toMinioErr(err, bucket, object, "")
-	}
-	log.Printf(
-		"bucket-name: %s, bucket-hash: %s, object-name: %s, object-hash: %s",
-		bucket, bucketHash, object, objectHash,
-	)
+	log.Printf("bucket-name: %s, object-name: %s", bucket, object)
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-object.go]: This should log the new bucket hash, and the object hash.  Basically anytime we log anything, we should always log the corresponding IPFS hashes if possible.

File: cmd/gateway/s3x/gateway-s3x.go [hunk 11]
```diff
@@ -91,6 +89,9 @@ func init() {
 			},
 		},
 	})
+	if err != nil {
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x.go]: should probably wrap in `if err := minio.RegisterGatewayCommand....` looks a bit cleaner

File: cmd/gateway/s3x/gateway-s3x-object.go [hunk 12]
```diff
@@ -176,42 +144,41 @@ func (x *xObjects) PutObject(
...
+
 	data, err := ioutil.ReadAll(r)
 	if err != nil {
-		return objInfo, x.toMinioErr(err, bucket, object, "")
+		return minio.ObjectInfo{}, x.toMinioErr(err, bucket, object, "")
 	}
-	obinfo.Size_ = int64(len(data))
-	dataHash, err := x.dagPut(ctx, data)
+	obinfo := newObjectInfo(bucket, object, len(data), opts)
+	dataHash, err := ipfsSaveBytes(ctx, x.dagClient, data)
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-object.go]: I made a mistake originally planning this. Object data itself needs to be stored as UnixFS data. Individual gRPC messages for TemporalX are capped at 4MB, so if the object data is larger than 4MB we will run into problems. Additionally storing object data as opposed to DAG's means the data can't be viewed on most gateways, whereas if we use UnixFS, the data can be viewed directly on IPFS gateways, etc... The `go-temporalx-client` package has helper `Upload` function that allows uploading a UnixFS file that's larger than 4MB without causing issues with gRPC max message sizes.

tl;dr - Any actual object data should now be stored as UnixFS on IPFS, and the MinIO protocol buffer objects (aka, the bucket buffer obj, and the object buffer obj) should continue to be stored as DAGs.

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+	}
+	return m, unlock, nil
+}
+
+// MultipartIDExists is used to lookup if the given multipart id exists
+func (ls *ledgerStore) MultipartIDExists(id string) error {
+	return ls.accertValidUploadID(id)
+}
+
+// GetMultipartHashes returns the hashes of all multipart upload object parts
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger-multipart.go:91]: this whole section should be removed yea?

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+		hashes[i] = objpart.GetDataHash()
+	}
+	return hashes, nil
+}*/
+
+///////////////////////
+// INTERNAL FUNCTINS //
+///////////////////////
+
+// accertValidUploadID is a helper function to check if a multipart id exists in our ledger
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger-multipart.go]: typo? should be `assertValidUploadID`  i think

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+	err := ls.ds.Delete(dsPartKey.ChildString(uploadID))
+	if err == datastore.ErrNotFound {
+		return ErrInvalidUploadID
+	}
+	return err
+}
+
+// getMultipartNilable returns a MultipartUpload or nil if it did not exist
+func (ls *ledgerStore) getMultipartNilable(uploadID string) (*MultipartUpload, error) {
+
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger-multipart.go]: nit pick, but remove this whitespace

File: cmd/gateway/s3x/gateway-s3x-ledger-store.go [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+func (ls *ledgerStore) object(ctx context.Context, bucket, object string) (*Object, error) {
+	h, err := ls.getObjectHash(ctx, bucket, object)
+	if err != nil {
+		return nil, err
+	}
+	return ipfsObject(ctx, ls.dag, h)
+}
+
+//ObjectInfo returns the ObjectInfo of the object.
+func (ls *ledgerStore) ObjectInfo(ctx context.Context, bucket, object string) (*ObjectInfo, error) {
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger-store.go:80]: Why use a pointer? If we don't use a pointer the code can be cleaned up to be:
```Golang
defer ls.locker.read(bucket)()
return ls.object(ctx, bucket, object)
```

File: cmd/gateway/s3x/gateway-s3x-ledger.go [hunk 17]
```diff
@@ -15,418 +17,84 @@ Any claiming or releasing of locks should be done in the public setter+getter fu
...
-	return le.putLedger(ledger)
-}
-
 // Close shuts down the ledger datastore
-func (le *LedgerStore) Close() error {
-	le.Lock()
-	defer le.Unlock()
-	return le.ds.Close()
+func (ls *ledgerStore) Close() error {
+	//todo: clean up caches
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger.go:22]: shouldn't we lock here to prevent access to the datastore during shutdown?

File: cmd/gateway/s3x/gateway-s3x-ledger.go [hunk 18]
```diff
@@ -15,418 +17,84 @@ Any claiming or releasing of locks should be done in the public setter+getter fu
...
+// GetObjectInfos returns a list of ordered ObjectInfos with given prefix ordered by name
+func (ls *ledgerStore) GetObjectInfos(ctx context.Context, bucket, prefix, startsFrom string, max int) ([]ObjectInfo, error) {
+	defer ls.locker.read(bucket)()
+	b, err := ls.getBucketLoaded(ctx, bucket)
 	if err != nil {
 		return nil, err
 	}
-	if err := le.multipartExists(ledger, id); err != nil {
-		return nil, err
+	var names []string
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger.go:37]: we should initialize this to `max` that way on large queries we have much better memory usage. Probably do something like:

`var names = make([]string, 0, max)`

File: cmd/gateway/s3x/gateway-s3x-ledger.go [hunk 19]
```diff
@@ -15,418 +17,84 @@ Any claiming or releasing of locks should be done in the public setter+getter fu
...
-// ObjectExists is a public function to check if an object exists, and returns the reason
-// the object can't be found if any
-func (le *LedgerStore) ObjectExists(bucketName, objectName string) error {
-	le.RLock()
-	defer le.RUnlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
+	list := make([]ObjectInfo, 0, len(names))
+	for _, name := range names {
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger.go:49]: should do `for i, name := range names`, that way we can rewrite like this to avoid usage of append

```Golang
for i, name := range names {
  obj, err := ls.object(ctx, bucket, name)
  if err != nil {
     return nil, err
  }
  list[i] = obj.GetObjectInfo()
)
```

File: cmd/gateway/s3x/gateway-s3x-ledger.go [hunk 20]
```diff
@@ -15,418 +17,84 @@ Any claiming or releasing of locks should be done in the public setter+getter fu
...
-}
-
-// GetBucketNames is used to a slice of all bucket names our ledger currently tracks
-func (le *LedgerStore) GetBucketNames() ([]string, error) {
-	le.RLock()
-	defer le.RUnlock()
-	ledger, err := le.getLedger()
+// GetBucketNames is used to get a slice of all bucket names our ledger currently tracks
+func (ls *ledgerStore) GetBucketNames() ([]string, error) {
+	//this only reads from the datastore, which have it's own synchronization, so no locking is needed.
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger.go:87]: smart

File: cmd/gateway/s3x/gateway-s3x-multipart.go [hunk 21]
```diff
@@ -107,17 +107,24 @@ func (x *xObjects) ListObjectParts(
...
+	if m.GetObjectInfo().GetBucket() != bucket ||
+		m.GetObjectInfo().GetName() != object {
+		return lpi, x.toMinioErr(ErrInvalidUploadID, bucket, object, uploadID)
+	}
+
+	for _, part := range m.ObjectParts {
 		lpi.Parts = append(lpi.Parts, minio.PartInfo{
 			PartNumber: int(part.GetNumber()),
-			ETag:       minio.ToS3ETag(part.GetEtag()),
+			ETag:       minio.ToS3ETag(part.GetDataHash()),
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-multipart.go:96]: I think this is a typo? It should be `part.GetETag` not `part.GetDataHash`

**Review by bonedaddy — CHANGES_REQUESTED** [review]
Couple of things to change, but by and large great fucking work.
Biggest thing to change is that actual "object data" should be stored as UnixFS IPLD types. See our discord PM's for more information

File: cmd/gateway/s3x/gateway-s3x-multipart.go (hunk 21 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-multipart.go:96]: I removed the Etag field and started using DataHash as the Etag

File: cmd/gateway/s3x/gateway-s3x-ledger.go (hunk 19 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger.go:49]: Personally, I like the append better, and with the capacity set, both code should have the same performance.

File: cmd/gateway/s3x/gateway-s3x-ledger.go (hunk 18 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger.go:37]: But it will take more memory for short queries.

File: cmd/gateway/s3x/gateway-s3x-ledger.go (hunk 17 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger.go:22]: Yes, I have been putting off this part. What if there are other API calls in progress? Also, there should be standard errors when calling the API after close.

File: cmd/gateway/s3x/gateway-s3x-ledger-store.go (hunk 16 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger-store.go:80]: Actually `ls.object` returns a point to `object`, not `ObjectInfo`. Otherwise, I want to avoid needlessly copying data.

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go (hunk 15 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger-multipart.go]: good catch

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go (hunk 14 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger-multipart.go]: thanks

File: cmd/gateway/s3x/gateway-s3x-ledger-store.go (hunk 16 above)
Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-ledger-store.go:80]: Good point

File: cmd/gateway/s3x/gateway-s3x-ledger.go (hunk 17 above)
Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-ledger.go:22]: That's a really good point. I'm honestly not sure what the answer is. I think for now just go with using a lock, and I'll create a GitHub issue to track research as to what the answer is. 

It will probably be something like prevent closing while there are active requests, but prevent new requests from being made

File: cmd/gateway/s3x/gateway-s3x-multipart.go (hunk 21 above)
Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-multipart.go:96]: makes sense

File: cmd/gateway/s3x/gateway-s3x-ledger.go (hunk 19 above)
Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-ledger.go:49]: good point

File: cmd/gateway/s3x/gateway-s3x-ledger.go (hunk 18 above)
Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-ledger.go:37]: good point

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go (hunk 13 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger-multipart.go:91]: Don't we need this for hash exporting API?

File: cmd/gateway/s3x/gateway-s3x.go (hunk 11 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x.go]: ok

File: cmd/gateway/s3x/gateway-s3x-object.go (hunk 10 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-object.go]: I removed it because retrieving it was tedious using the new code, I can add it back if necessary.

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go (hunk 9 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger-multipart.go:27]: At the end of gateway-s3x-multipart.go, still not implemented.

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go (hunk 13 above)
Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-ledger-multipart.go:91]: Actually I'm not sure, since multipart isn't done yet let's keep this here and decide if this should go when we finish multipart

File: cmd/gateway/s3x/gateway-s3x-object.go (hunk 10 above)
Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-object.go]: Yea I can see that. leave it out then.

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go (hunk 9 above)
Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-ledger-multipart.go:27]: right

File: cmd/gateway/s3x/gateway-s3x-object.go (hunk 12 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-object.go]: This change has just been made, please check

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+	if err != nil {
+		return err
+	}
+	return ls.DeleteMultipartID(multipartID)
+}
+
+// NewMultipartUpload is used to store the initial start of a multipart upload request
+func (ls *ledgerStore) NewMultipartUpload(multipartID string, info *ObjectInfo) error {
+	bucket := info.GetBucket()
+	defer ls.locker.write(bucket)
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger-multipart.go]: shouldn't this be declared after the `err := ls.assertBucketExists`, since if the bucket doesnt' exist there is no point to trigger a lock write?

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+	delete(ls.l.MultipartUploads, uploadID)
+	err := ls.ds.Delete(dsPartKey.ChildString(uploadID))
+	if err == datastore.ErrNotFound {
+		return ErrInvalidUploadID
+	}
+	return err
+}
+
+// getMultipartNilable returns a MultipartUpload or nil if it did not exist
+func (ls *ledgerStore) getMultipartNilable(uploadID string) (*MultipartUpload, error) {
```
Comment (bonedaddy) [inline cmd/gateway/s3x/gateway-s3x-ledger-multipart.go:145]: if we change the function signature to

```Go
func (ls *ledgerStore) getMultipartNilable(uploadID string) (mu *MultipartUpload, err error)
```

then the function can be condensed a bit to

```Go
// getMultipartNilable returns a MultipartUpload or nil if it did not exist
func (ls *ledgerStore) getMultipartNilable(uploadID string) (mu *MultipartUpload, err error) {
        ls.pmapLocker.Lock()
        defer ls.pmapLocker.Unlock()
        mu, ok := ls.l.MultipartUploads[uploadID]
        if !ok {
                data, err := ls.ds.Get(dsPartKey.ChildString(uploadID))
                if err == datastore.ErrNotFound {
                        return nil, nil //not found is nil, nil as documented
                }
                if err != nil {
                        return nil, err
                }
                mu = &MultipartUpload{}
                err = mu.Unmarshal(data)
        }
        return mu, err
}
```

Not a "requested change" ultimately just a personal preference but you can decide if its a change worth doing.

File: cmd/gateway/s3x/gateway-s3x-object.go (hunk 12 above)
Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-object.go]: Hell yea, nice job.

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go (hunk 23 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger-multipart.go:145]: `data, err := ls.ds.Get(dsPartKey.ChildString(uploadID))`
I think this line could shadow error.

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go (hunk 22 above)
Comment (xiegeo) [reply cmd/gateway/s3x/gateway-s3x-ledger-multipart.go]: Actually, it's not needed at all. Since there are no changes to bucket objects.

File: cmd/gateway/s3x/gateway-s3x-ledger-multipart.go (hunk 23 above)
Comment (bonedaddy) [reply cmd/gateway/s3x/gateway-s3x-ledger-multipart.go:145]: Good point, i like your changes here https://github.com/RTradeLtd/s3x/pull/16/commits/507c8b8f6a392dc23b247d7ac5ce8627d0261577 more so ill resolve this.

**Review by bonedaddy — APPROVED** [review]
Excellent work, after this I think it's best if you start on the multipart upload.
